File: Dot examples/ASP Book Examples/Book Examples/Chapter18/CS/LanguageChanges/App_Code/GenericMethods.cs
using System;
using System.Collections.Generic;
using System.Data.SqlClient;

public class GenericMethods
{
    public static List<T> GetListFromCommand<T>(SqlCommand command)
    where T : ICreatable, new()
    {
        List<T> results = new List<T>();
        using (command.Connection)
        {
            command.Connection.Open();
            SqlDataReader reader = command.ExecuteReader();
            while (reader.Read())
            {
                T newThing = new T();
                newThing.Create(reader);
                results.Add(newThing);
            }
        }
        return results;
    }
}

public interface ICreatable
{
    void Create(SqlDataReader reader);
}

File: Dot examples/ASP Book Examples/Book Examples/Chapter18/VB/LanguageChanges/App_Code/GenericMethods.vb
Imports System
Imports System.Collections.Generic
Imports System.Data.SqlClient

Public Class GenericMethods

    Public Shared Function GetListFromCommand(Of T As {ICreatable, New})(ByVal command As SqlCommand) As List(Of T)
        Dim results As New List(Of T)()
        Using command.Connection
            command.Connection.Open()
            Dim reader As SqlDataReader = command.ExecuteReader()
            While (reader.Read())
                Dim newThing As New T()
                newThing.Create(reader)
                results.Add(newThing)
            End While
        End Using
        Return results
    End Function
End Class

Public Interface ICreatable
    Sub Create(ByVal reader As SqlDataReader)
End Interface
